presto-raptor/src/main/java/com/facebook/presto/raptor/backup/BackgroundShardBackupManager.java

```diff
@@ -0,0 +1,142 @@
+import static com.google.common.base.Verify.verify;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.airlift.units.DataSize.Unit.BYTE;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+
+public class BackgroundShardBackupManager
+{
+    private final Optional<BackupStore> backupStore;
+    private final ExecutorService executorService;
+
+    private final AtomicInteger pendingBackups = new AtomicInteger();
+    private final BackupStats stats;
+
+    @Inject
+    public BackgroundShardBackupManager(Optional<BackupStore> backupStore, BackupConfig config)
+    {
+        this(backupStore, config.getBackgroundBackupThreads());
+    }
+
+    public BackgroundShardBackupManager(Optional<BackupStore> backupStore, int backupThreads)
+    {
+        checkArgument(backupThreads > 0, "backupThreads must be > 0");
+
+        this.backupStore = requireNonNull(backupStore, "backupStore is null");
+        this.executorService = newFixedThreadPool(backupThreads, daemonThreadsNamed("background-shard-backup-%s"));
+        this.stats = new BackupStats();
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        executorService.shutdownNow();
+    }
+
+    public CompletableFuture<Void> submit(UUID uuid, File source)
+    {
+        requireNonNull(uuid, "uuid is null");
+        requireNonNull(source, "source is null");
+
+        if (!backupStore.isPresent()) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        BackgroundBackup job = new BackgroundBackup(uuid, source);
+        executorService.submit(job);
+
+        pendingBackups.incrementAndGet();
+        job.getFuture().whenComplete((result, ex) -> pendingBackups.decrementAndGet());
+
+        return job.getFuture();
+    }
+
+    private class BackgroundBackup
+            implements Callable<Void>
+    {
+        private final UUID uuid;
+        private final File source;
+        private final long queuedTime;
+        private final CompletableFuture<Void> future;
+
+        public BackgroundBackup(UUID uuid, File source)
+        {
+            this.uuid = requireNonNull(uuid, "uuid is null");
+            this.source = requireNonNull(source, "source is null");
+
+            this.future = new CompletableFuture<>();
+            this.queuedTime = System.nanoTime();
+        }
+
+        @Override
+        public Void call()
+        {
+            try {
+                long start = System.nanoTime();
+
+                backupStore.get().backupShard(uuid, source);
+                verify(future.complete(null), "Failed to complete future");
+
+                stats.addCopyShardDataRate(new DataSize(source.length(), BYTE), Duration.nanosSince(start));
+                stats.addQueuedTime(Duration.nanosSince(queuedTime));
+                stats.incrementBackupSuccess();
+            }
+            catch (Exception exception) {
```

Comment:
Catch `Throwable` here, otherwise it will be lost in the `Executor` and the future will never complete